core/src/main/java/edu/wpi/grip/core/Pipeline.java

```diff
@@ -151,20 +249,24 @@ public void onSourceRemoved(SourceRemovedEvent event) {
         }
     }
 
-    public synchronized void addStep(int index, Step step) {
+    public void addStep(int index, Step step) {
```

Comment:
> If there is an exception and you forget to unlock in a finally block you could deadlock the entire application.

```java
lock.lock();
try {
    eventBus.register(step);
    eventBus.post(new StepAddedEvent(step, index));
} finally {
    lock.unlock();
}
```

> It ensures you lock an unlock correctly

I'm not sure I understand this.  If you do the above, the lock will definitely get unlocked no matter what.  Since it's private, you don't have to worry about any external class messing with it, and someone reading this method can clearly see exactly how it works.

I believe this is pretty standard usage of `Lock`, and I'm not sure why we would abstract that into something else unless it were for public methods.